rq/job.py

```diff
@@ -1679,8 +1695,77 @@ def __init__(self, max: int, interval: Union[int, Iterable[int]] = 0):
         self.intervals = intervals
 
 
+class CallbackType(str, Enum):
+    SUCCESS = 'success'
+    FAILURE = 'failure'
+    STOPPED = 'stopped'
+
+
+@runtime_checkable
+class SuccessCallbackFunc(Protocol):
+    def __call__(self, job: Job, connection: Redis, result: Any, *args: Any, **kwargs: Any) -> Any:
+        '''
+        Callback function to be called when a job is successfully executed.
+
+        Args:
+            job (Job): the job that was executed
+            connection (Redis): the Redis connection
+            result (Any): the result of the job execution
+            *args (Any): additional positional arguments
+            **kwargs (Any): additional keyword arguments
+        '''
+
+
+@runtime_checkable
+class FailureCallbackFunc(Protocol):
+    def __call__(
+        self, job: Job, connection: Redis, type: Type[BaseException], value: BaseException, traceback: Any
+    ) -> Any:
+        '''
+        Callback function to be called when a job fails during execution.
+
+        Args:
+            job (Job): the job that was executed
+            connection (Redis): the Redis connection
+            Type (type[BaseException]): the type of the exception
+            value (BaseException): the exception instance
+            traceback (Any): the traceback of the exception
+        '''
+
+
+@runtime_checkable
+class StoppedCallbackFunc(Protocol):
+    def __call__(self, job: Job, connection: Redis) -> Any:
+        '''
+        Callback function to be called when a job is stopped before execution.
+
+        Args:
+            job (Job): the job that was executed
+            connection (Redis): the Redis connection
+        '''
+
+
+CALLBACK_PROTOCOLS = {
+    CallbackType.SUCCESS: SuccessCallbackFunc,
+    CallbackType.FAILURE: FailureCallbackFunc,
+    CallbackType.STOPPED: StoppedCallbackFunc,
+}
+
+EXPECTED_SIGNATURE = {
+    CallbackType.SUCCESS: '(job: Job, connection: Redis, result: Any, *args: Any, **kwargs: Any) -> Any',
+    CallbackType.FAILURE: (
+        '(job: Job, connection: Redis, type: Type[BaseException], ' 'value: BaseException, traceback: Any) -> Any'
+    ),
+    CallbackType.STOPPED: '(job: Job, connection: Redis) -> Any',
+}
+
+
 class Callback:
-    def __init__(self, func: Union[str, Callable[..., Any]], timeout: Optional[Any] = None):
+    def __init__(
+        self,
+        func: Union[str, SuccessCallbackFunc, FailureCallbackFunc, StoppedCallbackFunc],
```

Comment:
if you make this generic this would look like:

```python
CallbackFn = TypeVar('CallbackFn', bound='Union[str, SuccessCallbackFunc, FailureCallbackFunc, StoppedCallbackFunc]')

class Callback(Generic[CallbackFn]):
    def __init__(self, func: CallbackFn, timeout: Optional[Any] = None) -> None:
```

and as mentioned above you could better type the `Job.create` for not only the legacy call-spec